Fix test for broken libjulia-codegen

apaz-cli opened the pull request:
Fails on windows due to utf-16 conversion.

Test was introduced on PR #47343, and has been talked about on #44527

apaz-cli committed:
Fix test for broken libjulia-codegen

Fails on windows due to utf-16 conversion.

Test was introduced on PR #47343

DilumAluthge commented:
The `test x86_64-w64-mingw32` job is showing:

```
Got exception outside of a @test
LoadError: UndefVarError: `readbytes` not defined
```

DilumAluthge commented:
I'm also not sure where `utf16` is defined.

DilumAluthge commented on test/compiler/codegen.jl:
```diff
@@ -707,7 +707,13 @@ mktempdir() do pfx
-    @test contains(readline(errfile), "ERROR: Unable to load dependent library")
+    errmsg = ""
+    @static if Sys.iswindows()
+        errmsg = open(errfile, "r") do f; utf16(readbytes(f)); end
```
```suggestion
        errmsg = open(errfile, "r") do f; utf16(read(f)); end
```

DilumAluthge committed:
Apply suggestions from code review

DilumAluthge commented on test/compiler/codegen.jl:
```diff
@@ -707,7 +707,13 @@ mktempdir() do pfx
     @test libs_emptied > 0
     @test_throws ProcessFailedException run(pipeline(`$pfx/bin/$(Base.julia_exename()) -e 'print("This should fail!\n")'`; stderr=errfile))
-    @test contains(readline(errfile), "ERROR: Unable to load dependent library")
+    errmsg = ""
```
```suggestion
    errmsg = ""
    utf16 = bytes -> begin
        n = Int(length(bytes)//2)
        pairs = [bytes[(2i-1):(2i)] for i in 1:n]
        f = x -> reinterpret(UInt16, x)
        units = only.(f.(pairs))
        return transcode(String, units)
    end
```

DilumAluthge committed:
Apply suggestions from code review

DilumAluthge commented on test/compiler/codegen.jl:
```diff
@@ -707,7 +707,20 @@ mktempdir() do pfx
+    utf16 = bytes -> begin
+        n = Int(length(bytes)//2)
+        pairs = [bytes[(2i-1):(2i)] for i in 1:n]
+        f = x -> reinterpret(UInt16, x)
```
```suggestion
        f = pair -> reinterpret(UInt16, pair)
```

DilumAluthge committed:
Apply suggestions from code review

DilumAluthge commented on test/compiler/codegen.jl:
```diff
@@ -707,7 +707,20 @@ mktempdir() do pfx
+        return transcode(String, units)
+    end
+    @static if Sys.iswindows()
+        errmsg = open(errfile, "r") do f; utf16(read(f)); end
```
```suggestion
        errmsg = utf16(read(errfile))
```

DilumAluthge committed:
Apply suggestions from code review

DilumAluthge commented on test/compiler/codegen.jl:
```diff
@@ -707,7 +707,20 @@ mktempdir() do pfx
     @test libs_emptied > 0
     @test_throws ProcessFailedException run(pipeline(`$pfx/bin/$(Base.julia_exename()) -e 'print("This should fail!\n")'`; stderr=errfile))
-    @test contains(readline(errfile), "ERROR: Unable to load dependent library")
+    errmsg = ""
```
```suggestion
```

DilumAluthge committed:
Apply suggestions from code review

giordano commented on test/compiler/codegen.jl:
```diff
@@ -707,7 +707,19 @@ mktempdir() do pfx
     @test libs_emptied > 0
     @test_throws ProcessFailedException run(pipeline(`$pfx/bin/$(Base.julia_exename()) -e 'print("This should fail!\n")'`; stderr=errfile))
-    @test contains(readline(errfile), "ERROR: Unable to load dependent library")
+    utf16 = bytes -> begin
+        n = Int(length(bytes)//2)
+        pairs = [bytes[(2i-1):(2i)] for i in 1:n]
+        f = pair -> reinterpret(UInt16, pair)
+        units = only.(f.(pairs))
+        return transcode(String, units)
+    end
+    @static if Sys.iswindows()
+        errmsg = utf16(read(errfile))
+    else
+        errmsg = readline(errfile)
+    end
```
Alternatively:
```julia
errmsg = replace(readline(errfile), "\0" => "")
```
? :grin:  It's less accurate because it blindly drops all `"\0"`, wherever they are, but it's also much simpler.

PallHaraldsson replied:
@apaz-cli FYI, I believe that's the right way (and much faster):
``` 
n = length(bytes) ÷ 2
```

See also div and \div TAB to type it. With rationals it's not also exactly the same, can throw, with odd number of bytes, which shouldn't happen... maybe you used it on purpose for that?

apaz-cli committed:
loader_fputs convert to utf16 only for win console

We use loader_fputs to write error messages, and when
we write them to files we want those files to be utf8.

This should fix a broken test introduced in #47343.

DilumAluthge commented:
@apaz-cli With https://github.com/JuliaLang/julia/pull/47404/commits/c8f15077af279787c14fbe9d64db736549e50146, I guess we can drop the changes to the test?

DilumAluthge commented:
Also, the PR title should be updated.

apaz-cli commented:
@DilumAluthge I talked about it with Jameson and Elliot, and the correct thing to do is to fix the write, rather than the read.

So, this PR should probably be thrown out and replaced with #47406

DilumAluthge commented:
Closing in favor of https://github.com/JuliaLang/julia/pull/47406
